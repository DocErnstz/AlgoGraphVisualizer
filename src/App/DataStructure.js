class QElement {
  constructor(element, priority) {
    this.element = element;
    this.priority = priority;
  }
}

class PriorityQueue {
  constructor() {
    this.items = [];
  }

  isEmpty() {
    return this.items.length == 0;
  }

  enqueue(element, priority) {
    var qElement = new QElement(element, priority);
    var contain = false;

    for (var i = 0; i < this.items.length; i++) {
      if (this.items[i].priority > qElement.priority) {
        this.items.splice(i, 0, qElement);
        contain = true;
        break;
      }
    }

    if (!contain) {
      this.items.push(qElement);
    }
  }

  dequeue() {
    if (this.isEmpty()) return "Underflow";
    return this.items.shift();
  }

  front() {
    if (this.isEmpty()) return "No elements in Queue";
    return this.items[0];
  }
}

class Cell {
  constructor(x, y) {
    this.x = x;
    this.y = y;
    this.id = x.toString() + y.toString();
  }
  neighbors() {
    var set = [];
    var minX = -1;
    var minY = -1;
    var maxX = 2;
    var maxY = 2;
    if (this.x - 1 < 0) minX = 0;
    if (this.x + 1 > 70) maxX = 1;
    if (this.y - 1 < 0) minY = 0;
    if (this.y + 1 > 31) maxY = 1;
    for (var x = minX; x < maxX; x++) {
      for (var y = minY; y < maxY; y++) {
        if (
          (x == 0 && y == 0) ||
          (x == 1 && y == 1) ||
          (x == -1 && y == -1) ||
          (x == -1 && y == 1) ||
          (x == 1 && y == -1)
        )
          continue;
        set.push(new Cell(x + this.x, y + this.y));
      }
    }

    return set;
  }
}

module.exports = { QElement, PriorityQueue, Cell };
