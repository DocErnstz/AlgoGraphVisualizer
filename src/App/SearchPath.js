const { PriorityQueue } = require("./DataStructure.js");
const { Cell } = require("./DataStructure.js");
import { blocks } from "./index.js";

export var cost_so_far = {};

export function select(x, y) {
  var elements = document.getElementById("matrix").children;
  const red = document.createElement("div");
  red.classList.add("red");
  elements.item(y).children.item(x).append(red);
}

export function Clear() {
  var elements = document.getElementById("matrix").children;
  cost_so_far = {};
  cost_so_far = Object.assign({}, blocks);

  for (var x = 0; x < 18; x++) {
    for (var y = 0; y < 18; y++) {
      if (elements.item(y).children.item(x).lastElementChild != null) {
        if (
          elements
            .item(y)
            .children.item(x)
            .lastElementChild.classList.contains("red")
        ) {
          elements.item(y).children.item(x).innerHTML = "";
        }
      }
    }
  }
}

export function ClearBlocks() {
  var elements = document.getElementById("matrix").children;
  cost_so_far = {};
  for (var x = 0; x < 18; x++) {
    for (var y = 0; y < 18; y++) {
      if (
        elements.item(y).children.item(x).className == "col block"
          ? true
          : false
      ) {
        elements.item(y).children.item(x).className = "col";
      }
    }
  }
}

export function UCS(start, goal) {
  var frontier = new PriorityQueue();
  frontier.enqueue(start, 0);
  var came_from = {};
  came_from[start.id] = null;
  cost_so_far[start.id] = 0;

  while (!frontier.isEmpty()) {
    var current = frontier.front();

    frontier.dequeue();
    if (current.element.id == goal.id) {
      break;
    }
    current.element.neighbors().forEach((next) => {
      var new_cost = 1 + cost_so_far[current.element.id];
      if (!(next.id in cost_so_far) || new_cost < cost_so_far[next.id]) {
        cost_so_far[next.id] = new_cost;
        frontier.enqueue(next, new_cost);
        came_from[next.id] = current.element;
        if (next.id != goal.id) {
          select(next.x, next.y);
        }
      }
    });
  }
  return came_from;
}

export function pathfinding(start, goal, UCS) {
  var path = [];
  var dict = UCS;
  var current = dict[goal.id];
  while (current.id != start.id) {
    path.push(current);
    current = dict[current.id];
  }
  return path;
}
