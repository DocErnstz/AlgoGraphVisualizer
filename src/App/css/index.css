@import url("https://fonts.googleapis.com/css2?family=Roboto:wght@300&display=swap");
@import "grid.css";
@import "utilities.css";
@import "navbar.css";

:root {
  --size_square: 17px;
  display: flex;
  justify-content: center;
  font-family: "Roboto", sans-serif;
}

body {
  margin: 0px;
  padding: 0px;
  width: 100%;
  height: 665px;
  display: flex;
  flex-direction: column;
  align-items: center;
}

#start,
#end {
  border-radius: 30px;
}

#start {
  background-color: red;
}
#end {
  background-color: blue;
}

body {
  background: white;
}
