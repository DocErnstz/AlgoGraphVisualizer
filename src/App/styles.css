
:root {
    --size_square: 30px;
    display: flex;
    justify-content: center;
}

body{
    margin: 30px;
    padding: 0px;
    height: 665px;
}

.row{
    display: flex;
    border: 1px solid black;
    width: calc(var(--size_square) * 18);
    height: var(--size_square);
}
.col{
 width: var(--size_square);
 height: var(--size_square);
 border: 1px solid black;
}

.red{
    background: red;
}