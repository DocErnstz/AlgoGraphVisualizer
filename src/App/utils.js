const { Cell } = require("./DataStructure.js");
export function block_setter(element) {
  element.classList.add("block");
  var block_point = new Cell(
    Array.from(element.parentNode.children).indexOf(element),
    Array.from(element.parentNode.parentNode.children).indexOf(
      element.parentNode
    )
  );
  return block_point;
}
