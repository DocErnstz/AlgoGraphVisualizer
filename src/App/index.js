require("./styles.css");

var elements=document.getElementById('matrix').children;

function select(x,y){
    elements.item(x).children.item(y).classList.add("red");
}




class Cell {
    constructor(x, y) {
      this.x = y;
      this.y = y;
    }
    neighbors(){
        var set = [];
        var minX = -1;
        var minY = -1;
        var maxX =  2;
        var maxY =  2;
        if (this.x-1 < 0)
          minX = 0;
        if (this.x+1 > 15)
          maxX = 1;
        if (this.y-1 < 0)
          minY = 0;
        if (this.y+1 > 15)
          maxY = 1;
        for (var x = minX; x < maxX; x++) {
          for (var y = minY; y < maxY; y++) {
            if((x == 0 && y == 0) || (x == 1 && y == 1) || (x == -1 && y == -1) || (x == -1 && y == 1) || (x == 1 && y == -1))
            continue;
            
            set.push(new Cell(x + this.x,y + this.y));
            
            
          }
        }
      return set
    }
  }

