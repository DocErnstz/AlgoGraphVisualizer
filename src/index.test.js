import { fireEvent, getByText, waitFor } from "@testing-library/dom";
import "@testing-library/jest-dom/extend-expect";
import { JSDOM } from "jsdom";
import fs from "fs";
import path from "path";
import { block_setter } from "./App/utils";

const html = fs.readFileSync(path.resolve(__dirname, "./index.html"), "utf8");

let dom;
let container;

describe("index.html", () => {
  beforeEach(() => {
    dom = new JSDOM(html, { runScripts: "dangerously" });
    container = dom.window.document.body;
  });

  it("renders a heading element", () => {
    expect(container.querySelector("h1")).not.toBeNull();
    expect(getByText(container, "Pathfinding")).toBeInTheDocument();
  });
  it("all cols are set up", async () => {
    const cols = container.getElementsByClassName("col");
    const colsArr = Array.from(cols);
    const colsArrF = colsArr.filter((x) => colsArr.indexOf(x) > 1);
    expect(colsArrF.length).toBe(322);
    expect(document.getElementsByClassName("col")).toBe([]);
  });
  it("block works out", () => {
    const scol = getByText(container, "a");
    const block_point = block_setter(scol);
    expect(scol).toHaveClass("block");
    const blockobject = {
      x: block_point.x,
      y: block_point.y,
      id: block_point.id,
    };
    expect(blockobject).toMatchSnapshot();
  });
});
