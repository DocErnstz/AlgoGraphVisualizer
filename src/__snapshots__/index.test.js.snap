// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index.html block works out 1`] = `
Object {
  "id": "30",
  "x": 3,
  "y": 0,
}
`;
